presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -367,9 +367,7 @@ public ActualProperties visitDistinctLimit(DistinctLimitNode node, List<ActualPr
         @Override
         public ActualProperties visitTableFinish(TableFinishNode node, List<ActualProperties> inputProperties)
         {
-            return ActualProperties.builder()
-                    .global(coordinatorSingleStreamPartition())
-                    .build();
+            return Iterables.getOnlyElement(inputProperties);
```

Comment:
I'm not sure this is correct.  The table finish operator instance (there is only one), gets the summary information from the `TableWriterOperator`s and then completes the table in the connector (e.g., in Hive adds the partitions to the pending transaction).  IIRC the output is just the number of rows written.  @electrum does that sound correct?

Reply:
You might be right. The code in `AddExchanges` for `TableFinishNode` is:
```
            ...
            if (!child.getProperties().isCoordinatorOnly()) {
                child = withDerivedProperties(
                        gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                        child.getProperties());
            }
            ...
```
so it creates single node partitioning if child is not running on coordinator.
I think that is should create coordinator partitioning instead.

Then we probably could remove `context.get().setCoordinatorOnlyDistribution();` from `Fragmenter#visitTableFinish`
In fact `AddExchanges.Rewriter#visitTableFinish` can probably be simplified to:
```
            PlanWithProperties child = planChild(node, PreferredProperties.any());
            if (!child.getProperties().isCoordinatorOnly()) {
                child = withDerivedProperties(
                        gatheringCoordinatorExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                        child.getProperties());
            }
            return rebaseAndDeriveProperties(node, child);
```

This way we
* create gathering coordinator exchange (to offload coordinator) even if child is running on a single (non-coordinator node) and
* don't create gathering exchange if stream is already running on coordinator.

This should decouple `AddExchanges` from `PlanFragmenter` a bit more

Reply:
Not sure I followed all of that, but the critical part is that `TableFinish` must run on the coordinator, so it can have direct access to the transaction state.